### PR TITLE
EZP-30969: Added method to fetch paginated list of content's reverse relations

### PR DESCRIPTION
**Pull request by mikadamczyk**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30969](https://jira.ez.no/browse/EZP-30969)
| **Bug/Improvement**| yes
| **Required by** | ezsystems/ezpublish-kernel#2774
| **New feature**    | no
| **Target version** | `7.5` for eZ Platform `v2.5.x` LTS
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

This method is needed to add pagination to content's reverse relation list. It is related to [https://jira.ez.no/browse/EZP-30900](https://jira.ez.no/browse/EZP-30900) and [https://jira.ez.no/browse/EZP-30968](https://jira.ez.no/browse/EZP-30968)

A command which can help during QA: https://gist.github.com/mikadamczyk/1d45ef638ffb18357883e7c301fa5c15

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by adamwojs** [comment]
@mikadamczyk Rebase is needed here.

File: eZ/Publish/API/Repository/Values/Content/RelationList/Item/RelationListItem.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        return $this->relation;
+    }
+
+    /**
+     * @return bool
+     */
+    public function hasRelation(): bool
+    {
+        return $this->relation instanceof Relation;
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Values/Content/RelationList/Item/RelationListItem.php]:
```suggestion
        return $this->relation;
```

Constructor guarantee that `$this->relation` will be always !== null

File: eZ/Publish/Core/Persistence/Legacy/Content/Gateway.php [hunk 2]
```diff
@@ -378,6 +378,18 @@ abstract class Gateway
...
+     * Loads paginated data of related to/from $contentId.
+     *
+     * @param int $contentId
+     * @param int $offset
+     * @param int $limit
+     * @param int $relationType
+     *
+     * @return string[][]
+     */
+    abstract public function listReverseRelations($contentId, int $offset = 0, int $limit = -1, $relationType = null): array;
```
Comment (adamwojs) [inline eZ/Publish/Core/Persistence/Legacy/Content/Gateway.php]:
```suggestion
    abstract public function listReverseRelations($contentId, int $offset = 0, int $limit = -1, ?int $relationType = null): array;
```

File: eZ/Publish/Core/Persistence/Legacy/Content/Gateway.php [hunk 3]
```diff
@@ -378,6 +378,18 @@ abstract class Gateway
      */
     abstract public function loadReverseRelations($contentId, $relationType = null);
 
+    /**
+     * Loads paginated data of related to/from $contentId.
+     *
+     * @param int $contentId
+     * @param int $offset
+     * @param int $limit
+     * @param int $relationType
```
Comment (adamwojs) [inline eZ/Publish/Core/Persistence/Legacy/Content/Gateway.php]:
```suggestion
     * @param ?int $relationType
```

File: eZ/Publish/Core/Persistence/Legacy/Content/Gateway.php [hunk 4]
```diff
@@ -378,6 +378,18 @@ abstract class Gateway
...
+     *
+     * @return string[][]
```
Comment (adamwojs) [inline eZ/Publish/Core/Persistence/Legacy/Content/Gateway.php]:
```suggestion
     * @return array
```

File: eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php [hunk 5]
```diff
@@ -1873,6 +1873,48 @@ public function loadReverseRelations($toContentId, $relationType = null)
...
+                $expr->eq('l.to_contentobject_id', ':toContentId')
+            )
+            ->innerJoin(
+                'l',
+                'ezcontentobject',
+                'c',
+                $expr->andX(
+                    $expr->eq('l.from_contentobject_id', 'c.id'),
+                    $expr->eq('l.from_contentobject_version', 'c.current_version'),
+                    $expr->eq('c.status', 1)
```
Comment (adamwojs) [inline eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php]: What's means status = `1`?

File: eZ/Publish/Core/Persistence/Legacy/Content/Gateway/ExceptionConversion.php [hunk 6]
```diff
@@ -714,6 +714,18 @@ public function loadReverseRelations($contentId, $relationType = null)
         }
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function listReverseRelations($contentId, int $offset = 0, int $limit = -1, ?int $relationType = null): array
```
Comment (alongosz) [inline eZ/Publish/Core/Persistence/Legacy/Content/Gateway/ExceptionConversion.php]:
```suggestion
    public function listReverseRelations(int $contentId, int $offset = 0, int $limit = -1, ?int $relationType = null): array
```

File: eZ/Publish/Core/Persistence/Cache/ContentHandler.php [hunk 7]
```diff
@@ -389,6 +389,25 @@ public function loadReverseRelations($destinationContentId, $type = null)
...
+     */
+    public function loadReverseRelationList(
+        int $destinationContentId,
+        int $offset = 0,
+        int $limit = -1,
+        ?int $type = null
+    ): array {
+        $this->logger->logCall(__METHOD__, ['content' => $destinationContentId, 'type' => $type]);
+
+        return $this->persistenceHandler->contentHandler()->loadReverseRelationList(
```
[the remark this answers is not in the document]
Comment (mikadamczyk) [reply eZ/Publish/Core/Persistence/Cache/ContentHandler.php]: As we discussed this will be done as follow-up

**Pull request merged**